internal/sys/ptr.go

```diff
@@ -1,43 +1,57 @@
 package sys
 
 import (
+	"math"
 	"unsafe"
 
 	"github.com/cilium/ebpf/internal/unix"
 )
 
-// NewPointer creates a 64-bit pointer from an unsafe Pointer.
-func NewPointer(ptr unsafe.Pointer) Pointer {
+// UnsafePointer creates a 64-bit pointer from an unsafe Pointer.
+func UnsafePointer(ptr unsafe.Pointer) Pointer {
 	return Pointer{ptr: ptr}
 }
 
-// NewSlicePointer creates a 64-bit pointer from a byte slice.
-func NewSlicePointer(buf []byte) Pointer {
-	if len(buf) == 0 {
+// SlicePointer creates a 64-bit pointer from a slice.
+func SlicePointer[E any](slice []E) Pointer {
+	if len(slice) == 0 {
 		return Pointer{}
 	}
 
-	return Pointer{ptr: unsafe.Pointer(&buf[0])}
+	return Pointer{ptr: unsafe.Pointer(&slice[0])}
 }
 
-// NewSlicePointer creates a 64-bit pointer from a byte slice.
+// SliceLen returns the length of a slice as a uint32.
 //
-// Useful to assign both the pointer and the length in one go.
-func NewSlicePointerLen(buf []byte) (Pointer, uint32) {
-	return NewSlicePointer(buf), uint32(len(buf))
+// Returns zero if the length of the slice exceeds uint32.
+func SliceLen[E any](slice []E) uint32 {
```

Comment:
This is only really useful if E is []byte.

Reply:
Maybe you meant `SliceElems`? This is useful with other types, check `QueryPrograms` in query.go that calls it with `[]ProgramID`.